### PR TITLE
UHM-5323: Lab Workflow lab results viewer should hide duplicative data

### PR DESCRIPTION
**Pull request by mogoodrich**

### CODE REVIEW

File: app/js/components/LabTrendsPage.jsx [hunk 1]
```diff
@@ -109,7 +110,7 @@ export class LabTrendsPage extends PureComponent {
 
     const formatChartData = data => data.map(item => ({
       ...item,
-      obsDatetime: moment(item.obsDatetime).format('DD-MMM-YYYY'),
+      obsDatetime: moment(item.obsDatetime).valueOf(),
```
Comment (mogoodrich) [inline app/js/components/LabTrendsPage.jsx:113]: This is related to fixing the the issue with x-axis of the trends graph, in UHM-5350

File: app/js/components/LabTrendsPage.jsx [hunk 2]
```diff
@@ -133,12 +134,18 @@ export class LabTrendsPage extends PureComponent {
       bottom: 5,
     };
 
-    const chartData = R.compose(
```
Comment (mogoodrich) [inline app/js/components/LabTrendsPage.jsx:136]: Compose processed right to left, which was confusing, I changed this to a pipe with operates left-to-right.

File: app/js/components/LabTrendsPage.jsx [hunk 3]
```diff
@@ -133,12 +134,18 @@ export class LabTrendsPage extends PureComponent {
...
+    const resultsToDisplay = R.pipe(
+      sortByDate('encounter.encounterDatetime'),
+      sortByDate('obsDatetime'),
+      filterDuplicates,
       R.reverse,
-      sortByDate('obsDateTime'),
```
Comment (mogoodrich) [inline app/js/components/LabTrendsPage.jsx:138]: There was a typo here in "obsDatetime".

**Review by cioan — APPROVED** [review]

File: app/js/utils/helpers.js [hunk 4]
```diff
@@ -77,6 +77,28 @@ export const sortByDate = (path) => data => R.sort(
   (a, b) => dateToInt(R.path(path.split('.'))(a)) - dateToInt(R.path(path.split('.'))(b)), data
 );
 
+// filters a list of obs, removing an obs if it has the same concept, value, and obsDatetime as the previous obs in the list
+export const filterDuplicates = data => data.reduce((acc, obs) => {
+  if (acc.length === 0) {
+    return [...acc, obs];
+  } else {
+    const previousObs = acc[acc.length - 1]
+    if (previousObs.obsDatetime !== obs.obsDatetime
```
Comment (mseaton) [inline app/js/utils/helpers.js:86]: From what I saw yesterday, there might be duplicates that are not immediately previous (i.e. multiple results on the same day of different values). Does this handle that case?

Comment (mogoodrich) [reply app/js/utils/helpers.js:86]: It does not... and I was reluctant to introduce that level of hackiness into it... I did, however, figure out why the sorting on the drill-down Lab Trends page wasn't working, so now at least the two pages are consistent.

Comment (mseaton) [reply app/js/utils/helpers.js:86]: Why is that hacky or hackier?

Comment (mogoodrich) [reply app/js/utils/helpers.js:86]: I guess it might not be hackier, but it would likely be a complete rewrite of this method, which I guess isn't the end of the world now that I've got some test cases...

Comment (mogoodrich) [reply app/js/utils/helpers.js]: @mseaton  I updated the PR to do what you asked about (remove duplicates that aren't immediately previous).  You can see the changes in the filterDuplicates method.  The main drawback is that this is now an N^2 magnitude operation (more or less) because each entry it tested against all entries before it to see if it's a duplicate.  Not super efficient, but it still seems to complete instantaneously so it didn't seem worth optimizing further (and potentially making it more confusing).  In particular, I prefer having all this logic in the filterDuplicates code rather than try to sort in a nonsensical manner (by value) just so all the duplicates end up next to each other.

Let me know if you have thoughts, will probably merge by end of day...  it *does* make the view for a patient with lots of duplicates look much better!

**Review by mseaton — APPROVED** [review]
